File: sorting_algorithms/bitonic_sort/bitonic_sort.js
/**
 * Bitonic sort is a comparison-based sorting algorithm that can be run in parallel.
 * It focuses on converting a random sequence of numbers into a bitonic sequence,
 * one that monotonically increases, then decreases.
 * The algorithm consists of two parts:
 *  1. Creating a bitonic sequence
 *  2. Sorting it
 * The first part is done recursively, and the second part is done iteratively.
 * Length of the input array must be a power of 2.
 * The algorithm is O(n log^2 n) in all cases.
 * Space complexity is O(n log^2 n) due to the recursive nature of the algorithm.
 */

/**
 * Bitonic merge subroutine.
 * @param {number[]} arr - array to be sorted
 * @param {number} lowIndex - starting index
 * @param {number} count - length of the array
 * @param {boolean} direction - direction of sorting, false for ascending, true for descending
 */
function bitonicMerge(arr, lowIndex, count, direction) {
  if (count <= 1) return;
  const length = count / 2;
  for (let i = lowIndex; i < lowIndex + length; i += 1) {
    if (arr[i] < arr[i + length] === direction) {
      [arr[i], arr[i + length]] = [arr[i + length], arr[i]];
    }
  }
  bitonicMerge(arr, lowIndex, length, direction);
  bitonicMerge(arr, lowIndex + length, length, direction);
}

/**
 * Bitonic sort algorithm.
 * @param {number[]} arr - array to be sorted
 * @param {number} lowIndex - starting index
 * @param {number} count - length of the array
 * @param {boolean} reversed - direction of sorting, true for ascending, false for descending
 */
function bitonicSort(arr, lowIndex, count, reversed = false) {
  if (count <= 1) return;
  const length = count / 2;

  // Split array into two parts
  bitonicSort(arr, lowIndex, length);
  bitonicSort(arr, lowIndex + length, length, true);

  // Merge the results
  bitonicMerge(arr, lowIndex, count, reversed);
}

/**
 * Sort function
 * @param {number[]} arr - array to be sorted
 * @param {boolean} reversed - direction of sorting
 */
function sort(arr, reversed = false) {
  bitonicSort(arr, 0, arr.length, reversed);
}

/**
 * Entry point.
 */
function main() {
  const arr = [1, 5, 2, 4, 3, 6, 8, 7];
  console.log('Array:', arr);
  sort(arr);
  console.log('Ascending bitonic sort:', arr);
  sort(arr, true);
  console.log('Descending bitonic sort:', arr);
}

main();

File: sorting_algorithms/bitonic_sort/bitonic_sort.py
#!/usr/bin/env python
"""
Bitonic sort is a comparison-based sorting algorithm that can be run in parallel.
It focuses on converting a random sequence of numbers into a bitonic sequence,
one that monotonically increases, then decreases.
The algorithm consists of two parts:
    1. Creating a bitonic sequence
    2. Sorting it
The first part is done recursively, and the second part is done iteratively.
Length of the input array must be a power of 2.
The algorithm is O(log^2 n) in all cases.
Space complexity is O(n) due to the recursive nature of the algorithm.
"""


def bitonic_merge(arr: list[int], low_index, count, direction: int):
    """
    Merge subroutine for bitonic sort.
    """
    if count > 1:
        k = count // 2
        # log(n) comparisons
        for i in range(low_index, low_index + k):
            if (arr[i] < arr[i + k]) == direction:
                arr[i], arr[i + k] = arr[i + k], arr[i]  # swap
        bitonic_merge(arr, low_index, k, direction)
        bitonic_merge(arr, low_index + k, k, direction)


def bitonic_sort(arr: list[int], low_index: int, count: int, direction: int):
    """
    Implements the bitonic sort algorithm.
    Parameters:
        arr: list of integers to be sorted
        low_index: starting index of the subarray
        count: number of elements to be sorted
        direction: 1 for ascending order, 0 for descending order
    """
    if count > 1:
        k = count // 2
        bitonic_sort(arr, low_index, k, 1)
        bitonic_sort(arr, low_index + k, k, 0)
        bitonic_merge(arr, low_index, count, direction)


def sort(arr: list[int], reverse: bool = False) -> list[int]:
    bitonic_sort(arr, 0, len(arr), reverse)
    return arr


if __name__ == '__main__':
    # arr = [3, 7, 4, 8, 6, 2, 1, 5]
    arr = [5, 10, 11, 11, 7, 2, 1, 3]
    print('Unsorted array:', arr)
    print('Ascending bitonic sort:', sort(arr))
    print('Descending bitonic sort:', sort(arr, reverse=True))
